mio/src/lib.rs

```diff
@@ -115,77 +149,128 @@ impl TcpStream {
-            }
-            Err(mio::channel::SendError::Io(e)) => {
-                return futures::failed(Error::new(e, into)).boxed()
-            }
-            Err(mio::channel::SendError::Disconnected(..)) => panic!("closed channel"),
-        }
     }
+}
+
+pub struct WriteCompletion {
+    tcp: Arc<mio::tcp::TcpStream>,
+    offset: usize,
+    data: Option<Vec<u8>>,
+}
 
-    pub fn write(&self, offset: usize, data: Vec<u8>)
-                 -> Box<Future<Item=(usize, Vec<u8>),
-                               Error=Error<(usize, Vec<u8>)>>> {
-        let r = (&*self.tcp).write(&data[offset..]);
+impl PollCompletion for WriteCompletion {
+    type Item = (usize, Vec<u8>);
+    type Error = Error<(usize, Vec<u8>)>;
+
+    fn poll_completion(&mut self) -> Option<PollResult<Self::Item, Self::Error>> {
+        let mut data = self.data.take().unwrap();
+        let r = (&*self.tcp).write(&data[self.offset..]);
         match r {
-            Ok(i) => return futures::finished((offset + i, data)).boxed(),
+            Ok(i) => Some(Ok((self.offset + i, data))),
             Err(e) => {
-                if e.kind() != io::ErrorKind::WouldBlock {
-                    return futures::failed(Error::new(e, (offset, data))).boxed()
+                if e.kind() == io::ErrorKind::WouldBlock {
+                    self.data = Some(data);
+                    None
+                } else {
+                    Some(Err(PollError::Other(Error::new(e, (self.offset, data)))))
                 }
             }
         }
-        let (p, c) = promise();
-        let r = self.tx.send(Message::Wait(c,
-                                           mio::EventSet::writable(),
-                                           self.tcp.clone()));
-        match r {
-            Ok(()) => {
-                let me2 = TcpStream {
-                    tcp: self.tcp.clone(),
-                    tx: self.tx.clone(),
-                };
-                p.then(move |res| {
-                    match res {
-                        Ok(()) => me2.write(offset, data),
-                        Err(e) => {
-                            futures::failed(Error::new(e, (offset, data))).boxed()
-                        }
-                    }
-                }).boxed()
-            }
-            Err(mio::channel::SendError::Io(e)) => {
-                return futures::failed(Error::new(e, (offset, data))).boxed()
-            }
-            Err(mio::channel::SendError::Disconnected(..)) => panic!("closed channel"),
+    }
+}
+
+unsafe fn slice_to_end(v: &mut Vec<u8>) -> &mut [u8] {
+    slice::from_raw_parts_mut(v.as_mut_ptr().offset(v.len() as isize),
+                              v.capacity() - v.len())
+}
+
+impl TcpStream {
+    pub fn local_addr(&self) -> io::Result<SocketAddr> {
+        self.tcp.local_addr()
+    }
+
+    pub fn peer_addr(&self) -> io::Result<SocketAddr> {
+        self.tcp.peer_addr()
+    }
+
+    pub fn ready_to_read(&self) -> MioEvent {
```

Comment:
I've wrestled in the past if we want an interface like this, and I think we may in the long run but we may want to keep it internal for awhile. Without a `Read` and `Write` implementation I'm not sure these'll be too useful though in isolation :(